### ISSUE TITLE
-fidasim_cyl flag crashes for W7-X restart run after depo

### ISSUE THREAD

**Issue by lazersos** [PrincetonUniversity/STELLOPT#200]
FIDASIM for a W7-X case crashes when using the -fidasim_cyl flag.  The -fidasim flag works fine.

The issue can be replicated by doing a restart from from a deposition run for W7-X.

**Comment by lazersos** [comment]
```
srun /u/sul/bin/xbeams3d -vmec W7X_KJM_n10_e30_i16_OPTEH2_birth -vessel /u/sul/pub/vessel/w7x/w7x_mod2_nbiport_highres_acc.dat -coil /u/sul/pub/coils/w7x/coils.w7x_mc_sc_tc_17_nfp5 -depo -fidasim_cyl -nobeamdensity
```

[input.W7X_KJM_n10_e30_i16_OPTEH2_birth.txt](https://github.com/PrincetonUniversity/STELLOPT/files/12643069/input.W7X_KJM_n10_e30_i16_OPTEH2_birth.txt)

[input.W7X_KJM_n10_e30_i16_OPTEH2_slow.txt](https://github.com/PrincetonUniversity/STELLOPT/files/12643071/input.W7X_KJM_n10_e30_i16_OPTEH2_slow.txt)

FAILS
```
srun /u/sul/bin/xbeams3d -vmec W7X_KJM_n10_e30_i16_OPTEH2_slow -restart beams3d_W7X_KJM_n10_e30_i16_OPTEH2_birth.h5 -plasma -collisions -fidasim_cyl -nobeamdensity
```

WORKS
```
srun /u/sul/bin/xbeams3d -vmec W7X_KJM_n10_e30_i16_OPTEH2_slow -restart beams3d_W7X_KJM_n10_e30_i16_OPTEH2_birth.h5 -plasma -collisions -fidasim -nobeamdensity
```

**Comment by kudav** [comment]
Where exactly does the execution fail? If its after the restart file message and before the vessel information, its likely an issue with running out of shared memory. I could run my cases by decreasing either the dist_prof or fida distribution resolution.

**Comment by lazersos** [comment]
Here's the output
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Temp Dir is:  /ptmp/sul/SLOW_OPTEMIST.7258942
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
srun /u/sul/bin/xbeams3d -vmec W7X_KJM_n10_e30_i16_OPTEH2_slow -restart beams3d_W7X_KJM_n10_e30_i16_OPTEH2_birth.h5 -plasma -collisions -fidasim_cyl -nobeamdensity
BEAMS3D Version  4.05
-----  HDF5 (Parallel) Parameters  -----
   HDF5_version:   1.12 release: 01
-----  MPI Parameters  -----
   MPI_version:   3.01
   Intel(R) MPI Library 2021.4 for Linux* OS

   Nproc_total:      4608
   Nproc_shared:       72
-----  GIT Repository  -----
   Repository: git@github.com:PrincetonUniversity/STELLOPT.git
   Branch:     feature/BEAMS3D_beamdensity
   Version:    v2.75-1910-g46f06-dirty
   Built-on:   17.09.2023 11:36:39
   Hash:       46f06c6f7aca501e0e83102072a78ef3d42ffed1
----- Input Parameters -----
   FILE: input.W7X_KJM_n10_e30_i16_OPTEH2_slow
   R   = [  4.25000,  6.50000];  NR:    128
   PHI = [ 0.00000, 1.25664];  NPHI:   72
   Z   = [-1.25000, 1.25000];  NZ:    128
   # of Particles to Start:        0
   COLLISION OPERATOR ON!
   FIDASIM OUTPUT ON!
   MAGNETIC FIELD FROM PLASMA ONLY!
   Restarting particles!
   RADIAL ELECTRIC FIELD PRESENT!
----- Plasma Parameters -----
   Te   = [  0.00000,  3.00000] keV;  NTE:     11;  S_MAX_TE:  1.00000
   Ti   = [  0.00000,  1.60000] keV;  NTI:     11;  S_MAX_TI:  1.00000
   Ne   = [  0.66667,  1.00000] E20 m^-3;  NNE:     11;  S_MAX_NE:  1.00000
   Ni(1)= [  0.66667,  1.00000] E20 m^-3;  M:   1 amu;  Z:  1;  S_MAX_NI:  1.00000
   Zeff = [  1.00000,  1.00000];  NZEFF:    6;  S_MAX_ZEFF:  1.00000
   V    = [ 12.76919, 15.52984] kV;  NPOT:   21;  S_MAX_POT:  1.00000
   PLASMA_MASS =    1.00784 amu
   PLASMA_ZMEAN =    1.00000 [Z]
----- VMEC Information -----
   FILE: W7X_KJM_n10_e30_i16_OPTEH2_slow
   R       = [  4.61016,  6.17646]
   Z       = [-0.89872, 0.89872]
   BETA    =    0.01;  I  =    1.38 [kA]
   AMINOR  =    0.50 [m]
   PHIEDGE =    1.73 [Wb]
   VOLUME  =   27.48 [m^3]
   CREATING WALL FROM HARMONICS
     Plasma Field Calculation [000]%[  0]%[  1]%[  2]%[  3]%[  3]%[  4]%[  5]%[  6]%[  7]%[  7]%[  8]%[  9]%[ 10]%[ 10]%[ 11]%[ 12]%[ 13]%[ 14]%[ 14]%[ 15]%[ 16]%[ 17]%[ 17]%[ 18]%[ 19]%[ 20]%[ 21]%[ 21]%[ 22]%[ 23]%[ 24]%[ 25]%[ 25]%[ 26]%[ 27]%[ 28]%[ 28]%[ 29]%[ 30]%[ 31]%[ 32]%[ 32]%[ 33]%[ 34]%[ 35]%[ 35]%[ 36]%[ 37]%[ 38]%[ 39]%[ 39]%[ 40]%[ 41]%[ 42]%[ 42]%[ 43]%[ 44]%[ 45]%[ 46]%[ 46]%[ 47]%[ 48]%[ 49]%[ 50]%[ 50]%[ 51]%[ 52]%[ 53]%[ 53]%[ 54]%[ 55]%[ 56]%[ 57]%[ 57]%[ 58]%[ 59]%[ 60]%[ 60]%[ 61]%[ 62]%[ 63]%[ 64]%[ 64]%[ 65]%[ 66]%[ 67]%[ 67]%[ 68]%[ 69]%[ 70]%[ 71]%[ 71]%[ 72]%[ 73]%[ 74]%[ 75]%[ 75]%[ 76]%[ 77]%[ 78]%[ 78]%[ 79]%[ 80]%[ 81]%[ 82]%[ 82]%[ 83]%[ 84]%[ 85]%[ 85]%[ 86]%[ 87]%[ 88]%[ 89]%[ 89]%[ 90]%[ 91]%[ 92]%[ 92]%[ 93]%[ 94]%[ 95]%[ 96]%[ 96]%[ 97]%[ 98]%[ 99]%[100]%
----- Constructing Splines -----
   R   = [  4.25000,  6.50000];  NR:    128
   PHI = [ 0.00000, 1.25664];  NPHI:   72
   Z   = [-1.25000, 1.25000];  NZ:    128
   HERMITE FORM: 1
----- Reading Restart File -----
   FILE: beams3d_W7X_KJM_n10_e30_i16_OPTEH2_birth.h5
   NPARTICLES_OLD:  1378944
   NPOINC_OLD:        3
   Detected old deposition run! 
   Detected old plasma only run! 
   NPARTICLES_RESTART:  1245353
   # of Beams:     21
 -----  Vessel Information  -----
   Wall Name :  HARMONICS
   Date      :  TODAY
   Faces     :   43200
   Blocks    :    4500
   Mean faces per block:    242.00
   Highest faces per block:    2254
   R_WALL   = [  4.61016,  6.17646]
   Z_WALL   = [ -0.89934,  0.89934]
 56.88889 91.69055 51.20000  0.08889 32.00000
----- FIDASIM Grid Parameters -----
   T_FIDA   =   0.00000
   R_FIDA   = [  4.25000,  6.50000];  NR:    128
   PHI_FIDA = [ 0.87270, 1.22170];  NPHI:   32
   Z_FIDA   = [-1.25000, 1.25000];  NZ:    128
   ENERGY_FIDA   = [ 5.62500, 354.37500];  NENERGY:     32
   PITCH_FIDA   = [-0.98438, 0.98438];  NPITCH:     64
----- FOLLOWING GYROCENTER TRAJECTORIES -----
       Method: LSODE
   Particles:   1245353
       Steps:   1000002   dt_min:   25.138E-09   dt_max:  100.000E-09
      NPOINC:         3
         Tol:    1.000E-09  Type: 10
     Trajectory Calculation [  0]%
 
 
  Copying files:
     From:  /ptmp/sul/SLOW_OPTEMIST.7258942
     To:    /raven/u/sul/Sims/W7X/BEAMS3D/OPTEMIST/FIDASIM
 
 
                                  
================================= 
Global information about the job: 
================================= 
  
Job owner: sul(20527)
Job name:  SLOW_OPTEMIST
Node list: ravc[3001,3008,3018,3020-3022,3043-3044,3047,3049,3057,3059,3061-3062,3124-3126,3262,3275,3282-3283,3285,3288,3307,3309,3329,3343,3359,3371-3400,3409,3426,3430,3432,3687-3688]
Job start: Sun Sep 17 13:00:13 CEST 2023
Job end:   Sun Sep 17 13:25:52 CEST 2023
Work dir:  /raven/u/sul/Sims/W7X/BEAMS3D/OPTEMIST/FIDASIM
Command:   /raven/u/sul/Sims/W7X/BEAMS3D/OPTEMIST/FIDASIM/SLURM.batch
  
  
  
==========================================================================================
Information on jobsteps (Note: MaxRSS/AveRSS is the maximum/average over all 
tasks of the per-task memory high-water marks; cf. "man sacct"): 
==========================================================================================
  
JobID            JobName NNodes NTasks  NCPUS       MaxRSS       AveRSS    Elapsed ExitCode
------------- ---------- ------ ------ ------ ------------ ------------ ---------- --------
7258942       SLOW_OPTEM     64          9216                             00:25:39      0:0
7258942.0       xbeams3d     64   4608   9216    10845.56M      502.58M   00:25:33      0:9
  
Maximum memory per node: 818.812699 GB (defined as MaxRSS*Ntasks/NNodes)
CPU utilization: 97.1 %
```
and some of the error file
```
munmap_chunk(): invalid pointer
srun: error: ravc3288: task 1604: Aborted
srun: launch/slurm: _step_signal: Terminating StepId=7258942.0
slurmstepd: error: *** STEP 7258942.0 ON ravc3001 CANCELLED AT 2023-09-17T13:25:19 ***
malloc(): invalid size (unsorted)
[ravc3049:133015:0:133015] Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
malloc(): mismatching next->prev_size (unsorted)
malloc(): invalid size (unsorted)
corrupted size vs. prev_size
forrtl: error (78): process killed (SIGTERM)
Image              PC                Routine            Line        Source             
libnetcdf.so.19.0  00001455B7566F8C  for__signal_handl     Unknown  Unknown
libpthread-2.31.s  00001455B7F04910  Unknown               Unknown  Unknown
xbeams3d           00000000006DF3E2  Unknown               Unknown  Unknown
xbeams3d           00000000004D5652  Unknown               Unknown  Unknown
xbeams3d           00000000008BA10F  Unknown               Unknown  Unknown
xbeams3d           0000000000852FB6  Unknown               Unknown  Unknown
xbeams3d           000000000051E20B  Unknown               Unknown  Unknown
xbeams3d           00000000004210AF  Unknown               Unknown  Unknown
xbeams3d           00000000004D5FA1  Unknown               Unknown  Unknown
xbeams3d           000000000040C522  Unknown               Unknown  Unknown
libc-2.31.so       00001455B3BE124D  __libc_start_main     Unknown  Unknown
xbeams3d           000000000040C43A  Unknown               Unknown  Unknown
forrtl: error (78): process killed (SIGTERM)
```

Looking back through it looks like the it happens after the paritcles are followed.

**Comment by kudav** [comment]
The segmentation fault when fidasim_cyl is used was not fixed through the pointer nullification. I investigated further and realized that in mpi_sharmem integers get multiplied to calculate the whindow_size:

```
 SUBROUTINE mpialloc_5d_dbl(array,n1,n2,n3,n4,n5,subid,mymaster,share_comm,win)
      ! Libraries
      USE MPI
      USE ISO_C_BINDING
      IMPLICIT NONE
      ! Arguments
      DOUBLE PRECISION, POINTER, INTENT(inout) :: array(:,:,:,:,:)
      INTEGER, INTENT(in) :: n1
      INTEGER, INTENT(in) :: n2
      INTEGER, INTENT(in) :: n3
      INTEGER, INTENT(in) :: n4
      INTEGER, INTENT(in) :: n5
      INTEGER, INTENT(in) :: subid
      INTEGER, INTENT(in) :: mymaster
      INTEGER, INTENT(in) :: share_comm
      INTEGER, INTENT(inout) :: win
      ! Variables
      INTEGER :: disp_unit, ier
      INTEGER :: array_shape(5)
      INTEGER(KIND=MPI_ADDRESS_KIND) :: window_size
      TYPE(C_PTR) :: baseptr
      ! Initialization
      ier = 0
      array_shape(1) = n1
      array_shape(2) = n2
      array_shape(3) = n3
      array_shape(4) = n4
      array_shape(5) = n5
      disp_unit = 1
      window_size = 0_MPI_ADDRESS_KIND
      **IF (subid == mymaster) window_size = INT(n1*n2*n3*n4*n5,MPI_ADDRESS_KIND)*8_MPI_ADDRESS_KIND**
      CALL MPI_WIN_ALLOCATE_SHARED(window_size, disp_unit, MPI_INFO_NULL, share_comm, baseptr, win ,ier)
      IF (subid /= mymaster) CALL MPI_WIN_SHARED_QUERY(win, 0, window_size, disp_unit, baseptr, ier)
      CALL C_F_POINTER(baseptr, array, array_shape)
      RETURN
      END SUBROUTINE mpialloc_5d_dbl
```
For 5 and 6 dimensional arrays, this can easily cause overflows as the typecasting is done only after the product is calculated. 
The proposed fixes are in commit b52d334c44bdf65ef4ac4108b9e3bf4f41c9f7cd (bugfix/LIBSTELL_MPI branch)

**Comment by lazersos** [comment]
Issue fixed.
